zephyr: Add zephyr board aliases

celinakalus opened the pull request:
Zephyr 3.7 introduces a new hardware model, where boards can now contain optional board qualifiers. Parts of this board description are now separated by slashes. This conflicts with the makemehappy target strings.

To solve this, add the possibility to assign aliases to Zephyr board targets which do not contain slashes. Any Zephyr board target can now be uniquely identified by an alias assigned to it by

ft commented:
Hey Celina! I've finally found some time to look at this. Luckily, Zephyr seems to allow for a deprecation period with the old names¹, so that we don't have to shoot from the hip. :)

With this we'd do this:

```yaml
zephyr-aliases:
  - alias: stm32h747i_disco_m7
    board-string: stm32h747i_disco/stm32h747xx/m7
  - alias: stm32h747i_disco_m4
    board-string: stm32h747i_disco/stm32h747xx/m4
```

…where instead of `board-string` it would also be possible to use keys like `board-name`, `revision`, and `soc` to specify the board an alias is targetting. I wonder if we need this level of specificity in `mmh`. At its core `mmh` is just a tool that helps with calling complicated `cmake` command lines. For this, the semantics of the different fields in Zephyr's new slashed naming scheme don't really matter. What matters is that we can't have slashes in most places we're using board names.

I was thinking, it should be enough to just have a straight mapping like this:

```yaml
zephyr-aliases:
  stm32h747i-disco-m7: stm32h747i_disco/stm32h747xx/m7
  stm32h747i-disco-m4: stm32h747i_disco/stm32h747xx/m4
```

I think that should make implementation straight forward too. In almost all places, `mmh` should use the alias. Only when constructing a `cmake` command line should the full name be resolved and used.

I can see these places where this should work:

- `mmh system build` (system build — this works)
- `mmh system list` (listing system build instances — also works)
- `mmh run` (module build — this doesn't work AFAICS)
- `mmh list` (listing module build instances — this seems to work, though)

We should also check if an alias name contains a slash and abort. Similarly, if a board name in a build specification contains a slash, we should abort and point the user to aliases to resolve the situation.

Another thing that needs improvement is, that like this a `system.yaml` without a `zephyr-aliases` section is invalid. But these mappings should be optional.

WDYT?

¹ It seems to automatically pick the new name for now:
```
CMake Warning at zephyr/kernel/cmake/modules/boards.cmake:110 (message):
  Deprecated BOARD=native_posix_64 specified, board automatically changed to:
  native_posix/native/64.
```

celinakalus commented:
If you think `mmh` does not need this complexity, then I agree we can just do it this shortly. But then also, we can just get rid of the `zephyr-aliases` key alltogether and just allow two ways to write boards:

```yaml
boards:
  - nucleo_f767zi
  - stm32h747i_disco_m7: stm32h747i_disco/stm32h747xx/m7
  - stm32h747i_disco_m4: stm32h747i_disco/stm32h747xx/m4
```

What do you think?

ft commented:
Huhm. That is not a bad idea. I like it.

ft commented:
I thought about this some more: I think I do prefer the `zephyr-aliases` approach better after all: With the new proposal, the `boards` spec becomes a list of `Union(String, Dictionary)`. This is already on the surface a complication, because it can now be either of two types. But that's not the main reason for my reconsideration. With this, it would be possible to write down dictionaries with more than one key, which I think is nonsensical (as in: It encodes nothing meaningful).

With having the `zephyr-aliases` section, there is a straight mapping from alias to full name, that can be consulted easily in the very few places that need it (namely when constructing `cmake` command lines for Zephyr builds).

I think that is more straight forward.

Also, aliases introduced like that can be used consistently in the same build spec, which I think is another big upside for the previous approach.

WDYT?

celinakalus commented:
I agree, I noticed very early as well that defining it in this way would lead to many small complications. I think collecting them centrally makes the most sense.

I have adjusted the code accordingly. Both system and module builds are now supported. Aliases are checked for illegal characters, including the forward slash. Board names are taken as-is if no alias of that name exists. I have also added aliases to the system example, where the disco board with its two processor cores is now in need of this feature.

ft commented:
Hey again! I tried this today (sorry for the delay) and it looks pretty spot on. So thanks again!

This is now part of `master`. I just squashed the first two commits into one (the first was the initial implementation, and the second is the final implementation), and rebased the branch onto `master`.
